traits/trait_list_object.py

```diff
@@ -90,108 +53,89 @@ def __init__(self, index=0, removed=None, added=None):
         self.added = added
 
 
-class TraitList(list):
-    """ A subclass of list that validates and notifies listeners of changes.
+def _normalize_index(index, length):
+    """ Normalize integer index to range 0 to len (inclusive). """
+    index = operator.index(index)
+    if index < 0:
+        return max(0, length + index)
+    else:
+        return min(length, index)
 
-    Parameters
-    ----------
-    value : list
-        The value for the list
-    validator : callable, optional
-        Called to validate items in the list.
-        With the signature::
 
-            validator(current_list, index, removed, added)
+def _normalize_slice(index, length):
+    """ Normalize slice start and stop to range 0 to len (inclusive). """
 
-        If the validator is not provided, all values added to the list
-        are considered to be valid.
+    # Do not normalize if step is negative.
+    if index.step is not None and index.step < 0:
+        return index
 
-    notifiers : list of callable, optional
-        A list of callables with the signature::
+    return slice(
+        _normalize_index(0 if index.start is None else index.start, length),
+        _normalize_index(length if index.stop is None else index.stop, length),
+        index.step,
+    )
 
-            notifier(trait_list, index, removed, added)
 
-        If a notifier is not provided, an empty list of notifiers is
-        used.
+# Default item validator for TraitList.
 
-    Attributes
-    ----------
-    validator : callable, optional
-        Called to validate items in the list
-        With the signature::
 
-            validator(current_list, index, removed, added)
+def accept_anything(item):
+    """
+    List item validator which accepts any item and returns it unaltered.
+    """
+    return item
 
-        If the validator is not provided, all values added to the list
-        are considered to be valid.
 
+class TraitList(list):
+    """ A subclass of list that validates and notifies listeners of changes.
+
+    Parameters
+    ----------
+    value : iterable
+        Iterable providing the items for the list
+    item_validator : callable, optional
+        Called to validate and/or transform items added to the list. The
+        callable should accept a single item from the list and return
+        the transformed item, raising TraitError for invalid items. If
```

Comment:
Is it necessary to ensure that the user defined validator raises a `TraitError` and not something else ? Do we need code to transform the error type if it doesn't?

Reply:
What would code to transform the error type look like? We definitely don't want to catch _all_ exceptions here, for the usual reasons (silencing genuine errors, making debugging harder). So would you target particular exception types? If so, which ones, and why?

For this sort of API, where you're delegating responsibility for validation to user code, you want to give that user code a well-defined way to say "nope, this item isn't valid", so that the "invalid" message can be safely distinguished from other messages, like coding errors.

And one clean way to do that is to provide a dedicated exception to use for that; in this case, `TraitError`.